proposals/ParameterUpdate.md

```diff
@@ -0,0 +1,687 @@
+* Nested parameter aggregates and parameterized models as parameters, which is important for layer-based high level APIs.
+* General optimizers that work with any floating-point parameter aggregate type.
+
+Examples:
+* [MNIST](https://github.com/tensorflow/swift-models/blob/master/MNIST/MNIST.swift): demonstrates `ParameterAggregate`.
+* [Autoencoder](https://github.com/tensorflow/swift-models/blob/master/Autoencoder/Autoencoder.swift): demonstrates `Parameterized`.
+* Optimizers are a work-in-progress, see a sample sketch in this [PR description](https://github.com/apple/swift/pull/18171#issue-203387822).
+
+Implementation:
+* [Part 1: ParameterAggregate protocol and synthesis](https://github.com/apple/swift/pull/18140)
+* [Part 2: Parameterized protocol and synthesis](https://github.com/apple/swift/pull/18171)
+
+---
+
+Two protocols have been added to the TensorFlow module: `ParameterAggregate` and `Parameterized`.
+
+The `ParameterAggregate` protocol represents an aggregate of parameters. Types that conform to `ParameterAggregate` must specify a `Parameter` associated type and an update method.
+
+```swift
+/// A type representing an aggregate of parameters.
+public protocol ParameterAggregate {
+  /// The parameter type.
+  associatedtype Parameter
+
+  /// Update parameters with their gradient values, using an updater function.
+  mutating func update(
+    withGradients gradients: Self,
+    _ updater: (inout Parameter, Parameter) -> Void
+  )
+}
+```
+
+For structs whose stored properties all have the same type, the compiler can synthesize the `ParameterAggregate` protocol requirements.
+
+Here's an example:
+
+```swift
+struct Parameters : ParameterAggregate {
+  // Note: all stored properties have the same type.
+  var w: Tensor<Float>
+  var b: Tensor<Float>
+
+  // Note: all code below is compiler synthesized.
+
+  // The `Parameter` associated type is synthesized to be the common stored
+  // property type.
+  // typealias Parameter = Tensor<Float>
+
+  // The synthesized `update` method applies the `updater` argument to each
+  // parameter-gradient pair.
+  // mutating func update(
+  //   withGradients gradients: Parameters,
+  //   _ updater: (inout Parameter, Parameter) -> Void
+  // ) {
+  //   updater(&w, gradients.w)
+  //   updater(&b, gradients.b)
+  // }
+}
+
+// Contrived parameter update example.
+var parameters = Parameters(w: [[1, 1], [1, 1]]), b: Tensor(1))
+let gradients = Parameters(w: [[0.5, 0.5], [0.5, 0.5]], b: Tensor(0.5))
+parameters.update(withGradients: gradients) { p, g in
+  p -= 0.1 * g
+}
+print(parameters)
+// Parameters(w: [[0.95, 0.95], [0.95, 0.95]], b: 0.95)
+```
+
+Model creators can define custom structs that conforms to `ParameterAggregate`, and use them in functions:
+
+```swift
+/// Parameters of an MNIST classifier.
+struct MNISTParameters : ParameterAggregate {
+    var w1 = Tensor<Float>(randomUniform: [784, 30])
```

Comment:
Indentation is inconsistent across the doc.

Reply:
Thanks. I'll use 2 space indentation in code blocks (rather than 4 spaces), to save space.

Reply:
Done in https://github.com/tensorflow/swift/pull/53/commits/08a21ceb187c45847b825e8aa5493500264775b3.